microsoft/testsuites/virtualclient/workload.py

```diff
@@ -0,0 +1,120 @@
+import re
+import uuid
+from typing import Any, Dict, List, Tuple, cast
+
+from assertpy import assert_that
+
+from lisa import (
+    Logger,
+    RemoteNode,
+    TestCaseMetadata,
+    TestSuite,
+    TestSuiteMetadata,
+    simple_requirement,
+)
+from lisa.features import Disk
+from lisa.operating_system import BSD
+from lisa.testsuite import TestResult
+from lisa.tools import VirtualClientTool
+
+
+@TestSuiteMetadata(
+    area="virtual_client",
+    category="performance",
+    description="""
+        This test suite runs the performance test cases with Virtual Client.
+    """,
+)
+class VirtualClient(TestSuite):
+    # [07/08/2024 13:58:15] Exit Code: 0\r\n
+    exit_code_pattern = re.compile(r"([\w\W]*?)Exit Code: (?P<exit_code>\d+)")
+
+    @TestCaseMetadata(
+        description="""
+            This test is to run redis workload testing with Virtual Client.
+        """,
+        priority=2,
+        requirement=simple_requirement(
+            unsupported_os=[BSD],
+            min_count=2,
+        ),
+    )
+    def perf_redis(self, log: Logger, result: TestResult) -> None:
+        self._work_load(result=result, profile_name="PERF-REDIS", system="Azure")
+
+    @TestCaseMetadata(
+        description="""
+            This test is to run PostgreSQL workload testing with Virtual Client.
+        """,
+        priority=3,
+        requirement=simple_requirement(
+            unsupported_os=[BSD],
+            min_count=2,
+        ),
+        timeout=3000,
+    )
+    def perf_postgresql(self, log: Logger, result: TestResult) -> None:
+        environment = result.environment
+        assert environment, "the environment should not be empty"
+        for node in environment.nodes.list():
+            disk = node.features[Disk]
+            disk.add_data_disk(count=1, size_in_gb=256)
+
+        self._work_load(
+            result=result,
+            profile_name="PERF-POSTGRESQL-HAMMERDB-TPCC",
+            system="Azure",
+            timeout=45,
+        )
+
+    def _work_load(
+        self, result: TestResult, profile_name: str, system: str, timeout: int = 10
+    ) -> None:
+        environment = result.environment
+        assert environment, "fail to get environment from testresult"
+        nodes: List[Dict[str, Any]] = [
+            {"role": "server", "node": cast(RemoteNode, environment.nodes[1])},
+            {"role": "client", "node": cast(RemoteNode, environment.nodes[0])},
+        ]
+
+        # Prepare data for layout generation
+        clients_data: List[Tuple[str, str, str]] = [
+            (node_info["role"], node_info["node"].internal_address, node_info["role"])
+            for node_info in nodes
+        ]
+
+        processes = {}
+        experiment_id = str(uuid.uuid4())
+        for node_info in nodes:
+            node = node_info["node"]
+            vc_tool = node.tools[VirtualClientTool]
```

Comment:
use run in parallel.

Reply:
The tool installation is still run separated, it's slow down the test progress.

Reply:
![image](https://github.com/user-attachments/assets/ba109110-122e-4c3c-8675-de0cde91755a)

Reply:
It can call `tools.get`, so the private function is not needed.

Reply:
updated